File: Algorithm_Course/Homework_2/smart_photographer.py
import numpy as np
import time

class Graph(): 
  
    def __init__(self, total_num): 

        # define some variables, 
        # like total number of array,
        # adjancy list of the elements that the current list connect to,
        # current max and final max
        self.total_num = total_num 
        self.adj = [[] for i in range(total_num)] 
        self.current_max = 1
        self.max = 1

        # Dp array  
        self.dp_path = [1] * (total_num)  

        # Visited array to know if the node  
        # has been visited previously or not  
        self.vis = [False] * (total_num) 

    # define edge and add to the list
    def add_edge(self, x, y): 
        
        self.adj[x].append(y) 

    # use recursive function to find the child of the child... of the child,
    # and find the maximum path
    def find_max_path(self,x,seen=None,path=None) :
        
        # Here are code for recursive code, but is not dynamic programming.
        # dynamic programming is more efficient than others
        ''' 
        count = self.childcount(self.adj[x])
        # print("current : ",x,"\n====")
        # print("It's childs : ",self.adj[x],"\n====")
        if count == 0 :
            if self.current_max > self.max :
                self.max = self.current_max
            # print("current max : ",self.current_max,"\n====")
            self.current_max = 1
        else :
            for i in (self.adj[x]) :
                self.current_max = self.current_max + 1
                self.find_max_path(i)
        if seen is None: seen = []
        if path is None: path = [x]
        seen.append(x)
        paths = []
        for current_node in self.adj[x]:
            if current_node not in seen:
                current_node_path = path + [current_node]
                paths.append(tuple(current_node_path))
                paths.extend(self.find_max_path(current_node, seen[:], current_node_path))
        '''

        # here use dynamic programming to record each node if they are visited, 
        # and also record the path maximun
        # ======================================================================
        # call dfs for every unvisited vertex  
        for i in range(0,self.total_num):   
            if not self.vis[i]:  
                self.dfs(i)  
        
        maximum = 0 
        
        # compare and find the maximum of all self.dp[i]  
        for i in range(0,self.total_num):   
            maximum = max(maximum, self.dp_path[i])  
        
        return maximum 

    # call dfs, deep-first search, which mentioned in class
    # here we use 'visit' and  'dp_path' to record information
    # avoid to do the same operation again, that's why dynamic program
    def dfs(self,current_node):  
   
        # set current node as visit  
        self.vis[current_node] = True 
        
        # traverse for all its children  
        for i in range(0, len(self.adj[current_node])):   
        
            # if not visited  
            if not self.vis[self.adj[current_node][i]]: 

                # run the dfs for its children
                self.dfs(self.adj[current_node][i])  
        
            # store the max of the paths  
            self.dp_path[current_node] = max(self.dp_path[current_node], 1 + self.dp_path[self.adj[current_node][i]]) 

def partition(arr,low,high): 

    # like the algrithm, we comare if the first 
    # number is bigger or smaller than the piviot
    # if current number is smaller, then
    # put the number to the front part
    # if bigger, don't move the number
    
    # this is the index number to make smaller element 
    # swtich to the front
    i = ( low-1 )  

    # we set the piviot as the last element in array
    pivot = arr[high]      
    
    # compare numbers using for loop
    for j in range(low , high): 
        
        # as we mentioned above,
        # compare the two number of the current one and the piviot
        if   arr[j] <= pivot: 
          
            i = i+1 
            arr[i],arr[j] = arr[j],arr[i] 
    
    # after seperate smaller part and bigger part
    # we move piviot to the middle
    arr[i+1],arr[high] = arr[high],arr[i+1] 

    # return the place that the piviot is
    return ( i+1 ) 
  
# Function to do Quick sort 
def quickSort(arr,low,high): 
    if low < high: 
        
        # like teacher's slide, partition by the piviot
        piviot = partition(arr,low,high) 

        # Separately sort elements before 
        # piviot place's number and after piviot place's numbe
        quickSort(arr, low, piviot-1) 
        quickSort(arr, piviot+1, high) 

# sort each
def sort_each(arr_1,n_total) :

    # sort each line of students
    for i in range(n_total) :
        n = len(arr_1[i]) 
        quickSort(arr_1[i],0,n-1) 

    return arr_1

# find adjancy matrix
def create_adjancy_table(n_total,arr_1,adjancy_table) :

    # create a n*n table
    for x in range(n_total) :
        for y in range(n_total) :

            if x != y :
                # if status = 1, the elements in x are all smaller than another one
                # if status = 0, the elements in x are not all smaller than another one
                status = check_status(arr_1[x],arr_1[y])

                if status == 1 :
                    adjancy_table[x][y] = 1
    
    return adjancy_table

# check status
def check_status(array_x,array_y) :

    # declare current status
    status = 1

    for i in range(len(array_x)) :
        # print(array_x[i],array_y[i])
        if status !=0 :
            if array_x[i] <  array_y[i] :
                status = status*1
            else :
                status = status*0
        else :
            break

    # print("status : ",status)
    # print("=====================")

    return status
  
if __name__ == "__main__" :

    # input data
    input_data = input("input numbers and students numbers : ")
    input_data_each = list(map(int, input_data.split()))
    n_total = input_data_each[0]
    n_students = input_data_each[1]

    arr_1 = list()

    # print("Initialized Matrix : ")
    # print("=======================")
    
    for i in range(n_total) :

        each_height = input("input student's height : ")
        arr_1.append(list(map(int, each_height.split())))

        # print("Finish Enter " + str(i) + " row !")
        # print("=================================")

    # print("Start Processing :\n=================================")
    
    # create a dictionary to count in_node and out_node
    # adjancy_table = np.zeros((n_total,n_total))

    # use list instead of numpy
    adjancy_table = [[0 for col in range(n_total)] for row in range(n_total)]

    # quick sort find sorting list
    arr_1 = sort_each(arr_1,n_total)
    # print(arr_1)
    
    # find adjancy matrix
    adjancy_table = create_adjancy_table(n_total,arr_1,adjancy_table) 
    # print(adjancy_table)

    # create graph
    G = Graph(n_total)

    # use adjancy_table to create graph
    for x in range(n_total) :
        for y in range(n_total) :
            if adjancy_table[x][y] == 1 :
                G.add_edge(x,y)

    # use root to find max root
    current_max = 0
    root_choice = list()

    # count zero
    for x in range(n_total) :

        # count zero to get root
        count_zero = 0
        for y in range(n_total) :
            
            if adjancy_table[y][x] == 0 :
                count_zero = count_zero + 1

        # get the start root, and find max path
        if count_zero == n_total :
            root_choice.append(x)

    print("Adjancy table : ",G.adj)
                
    # go through each possible root
    max_len = 0
    for i in range(len(root_choice)) :
        max_len = G.find_max_path(root_choice[i])
        # if all_path :
        #     max_len = max(len(p) for p in all_path)
        if max_len > current_max :
            current_max = max_len

    # print result
    print("Maximum Grades is : ",current_max)

File: Algorithm_Course/Homework_2/smart_photographer.c
# include<stdio.h>
# include<stdlib.h>
# include<string.h>
# include<stdbool.h>

// declare AdjListNode and using pointer to delcare AdjListNode ( point to next )
struct AdjListNode{ 
    int dest; 
    struct AdjListNode* next; 
}; 

// declare Parent_node
struct Parent_node{
	int dfs_path; 
	int dfs_count;
	bool visited;
};

// declare AdjList
struct AdjList{ 
    struct AdjListNode *head;  
}; 

// declare struct graph and also the pointer AdjList and Parent_node
// which is using to create dynamic array
struct Graph{ 
    int total_num; 
    struct AdjList* array_adj; 
	struct Parent_node* array_parent; 
}; 

// declare struct dynamic new node list 
struct AdjListNode* newAdjListNode(int dest){ 
	struct AdjListNode* newNode = 
	(struct AdjListNode*) malloc(sizeof(struct AdjListNode)); 
	newNode -> dest = dest; 
	newNode -> next = NULL; 
	return newNode; 
} 

// create struct dynamic Grpah using struct "createGraph"
struct Graph* createGraph(int total_num){ 
	struct Graph* graph = (struct Graph*) malloc(sizeof(struct Graph)); 

	graph -> total_num = total_num; 
	graph -> array_adj = (struct AdjList*) malloc(total_num * sizeof(struct AdjList)); 
	graph -> array_parent = (struct Parent_node*) malloc(total_num * sizeof(struct Parent_node)); 

	for (int i=0; i < total_num; ++i){
		graph -> array_adj[i].head = NULL;
		graph -> array_parent[i].dfs_path = 1;
		graph -> array_parent[i].dfs_count = 0;
		graph -> array_parent[i].visited = false;
	}

	return graph; 
} 

// create a edge by adjancy matrix using exsited graph
void addEdge(struct Graph* graph, int src, int dest){ 
	// Add an edge from src to dest. A new node is 
	// added to the adjacency list of src. The node 
	// is added at the begining 
	struct AdjListNode* newNode = newAdjListNode(dest); 
	newNode -> next = graph -> array_adj[src].head; 
	graph -> array_adj[src].head = newNode; 
	graph -> array_parent[src].dfs_count = graph -> array_parent[src].dfs_count + 1;
	// printf("DEST : %d",newNode->dest);
} 

// this is the dfs function to get current max path number
void dfs(struct Graph* graph, struct AdjListNode* next,int mother){ 
	
	//  if next is not null 
	while(next){

		// set current node as visit  
		int current_smaller_index = next -> dest;
		// printf("%d\n====================\n",current_parents_number);
		graph -> array_parent[mother].visited = true;
		
		// check if the smaller one that we didn't go through ( if not visited )
		if(graph -> array_parent[current_smaller_index].visited!=true){

			struct AdjListNode* smaller_one = graph -> array_adj[current_smaller_index].head;
			
			// run the dfs for its the grade that smaller than before
			dfs(graph,smaller_one,current_smaller_index) ;
		}

		// store the max of the paths  
		if(graph -> array_parent[mother].dfs_path <= graph -> array_parent[current_smaller_index].dfs_path + 1 ){
			graph -> array_parent[mother].dfs_path = graph -> array_parent[current_smaller_index].dfs_path + 1;
		}else{
			graph -> array_parent[mother].dfs_path = graph -> array_parent[mother].dfs_path;
		}	

		// to the next one
		next = next -> next;
	}
	
}

// function to fin max path
int find_max_path(struct Graph* graph, int total_num){

	// declare the variables
	int maximum = 0 ;

	// for loop to check each node
	for(int i=0;i<total_num;i++){
		struct AdjListNode* parents = graph -> array_adj[i].head; 

		// if not null, run dfs
		if(parents == NULL){
			graph -> array_parent[i].visited = true;
		}else{
			if(graph->array_parent[i].visited!=true){
				dfs(graph,parents,i);
			}
		}
	}   
    
	// compare and find the maximum of all self.dp[i]  
    for(int i=0;i<total_num;i++){
		if(graph->array_parent[i].dfs_path>=maximum){
			maximum = graph->array_parent[i].dfs_path;
		}
	}

    return maximum ;
        
}
int partition(int *all_students,int low,int high){ 

    // like the algrithm, we comare if the first 
    // number is bigger or smaller than the piviot
    // if current number is smaller, then
    // put the number to the front part
    // if bigger, don't move the number
    
    // this is the index number to make smaller element 
    // swtich to the front
    int i = ( low-1 ) ;

    // we set the piviot as the last element in array
    int pivot = *(all_students+high) ; 

	// declare replacement
	int rep_first = 0;
    int rep_second = 0;

    // compare numbers using for loop
    for(int j=low; j<=high-1 ; j++){
        
        // as we mentioned above,
        // compare the two number of the current one and the piviot
        if(*(all_students+j) <= pivot){ 
          
            i = i+1 ;

            rep_first = *(all_students+i) ; 
			*(all_students+i) = *(all_students+j) ; 
			*(all_students+j) = rep_first ; 
		}
	}

	// after seperate smaller part and bigger part
	// we move piviot to the middle
	rep_second = *(all_students+i+1) ; 
	*(all_students+i+1) = *(all_students+high) ; 
	*(all_students+high) = rep_second ; 

    // return the place that the piviot is
    return ( i+1 );
}

// Function to do Quick sort 
void quickSort(int *all_students,int low,int high){
	
	int piviot;

    if(low < high){ 
        
        // like teacher's slide, partition by the piviot
        piviot = partition(all_students, low, high) ;

        // Separately sort elements before 
        // piviot place's number and after piviot place's numbe
        quickSort(all_students, low, piviot-1)  ;
        quickSort(all_students, piviot+1, high)  ;
	}
}

// check status
int check_status(int *all_students,int x,int y,int m){

    //declare current status
    int status = 1;

    for(int i=0;i<m;i++){
        int current_a = *(all_students+x*m+i);
		int current_b = *(all_students+y*m+i);
		// printf("\ncurrent x : %d\n",current_a);
		// printf("current y : %d\n===========",current_b);
        if(status !=0 ){
            if(current_a < current_b){
                status = status*1;
			}else{
                status = status*0;
			}
		}else{
            break;
		}
	}

    return status;
}

// find adjancy matrix
int create_adjancy_table(int m, int n,int *all_students,int adjancy_table[n][n]){
	// declare status
	int status = 0;

	// create a n*n table
    for(int x=0;x<n;x++){
		for(int y=0;y<n;y++){

            if(x != y ){
				
				status = 0;
                // if status = 1, the elements in x are all smaller than another one
                // if status = 0, the elements in x are not all smaller than another one
                status = check_status(all_students,x,y,m);

                if(status == 1 ){
                    adjancy_table[x][y] = 1;
					// printf("Hi");
				}
			}
		}
	}

	return adjancy_table[n][n];
}

// main function to run
int main(){

	// declare some variables
	int n = 0 , m = 0 , maximum = 0;
	
	scanf("%d %d",&n,&m);
	// int *record = (int*) malloc(sizeof(int) * n*m);

	// use pointer to declare 2d array by using one pointer to create n*m array
	int *all_students = (int*) malloc(sizeof(int) * n*m);
	int per_col = m-1;

	// use memset to dynamic declare 2d array for adjancy table
	int adjancy_table[n][n];
	memset( adjancy_table, 0, n*n*sizeof(int) );

	// use memset to dynamic declare 2d array for edge
	int edge[n];
	memset( edge, 0, n*sizeof(int) );
	
	// enter  data
	for(int i=0 ; i<n*m ; i++){
		scanf("%d", all_students+i);
	}

	// seperate from input and results
	// printf("\n=========================\n");

	// quick sort for each grade's students
	for(int i=0 ; i<n ; i++){
		quickSort(all_students,i*m,i*m+m-1);
	}

	// for(int i=0;i<n;i++){
	// 	for(int j=0;j<m;j++){
	// 		printf("%d,",*(all_students+i*m+j));
	// 	}
	// 	printf("\n===============\n");
	// }

	// printf("=====================\n");

	// create adjancy table dynamically
	adjancy_table[n][n] = create_adjancy_table(m,n,all_students,adjancy_table);

	// print adjancy table for checking
	// for(int i=0;i<n;i++){
	// 	for(int j=0;j<n;j++){
	// 		printf("%d,",adjancy_table[i][j]);
	// 	}
	// }

	// seperate from input and results
	// printf("=========================");

	// create graph for each grade's student
	struct Graph* graph = createGraph(n); 

	// use adjancy_table to create graph
    for(int x=0;x<n;x++){
		for(int y=0;y<n;y++){
			if(adjancy_table[x][y] == 1){
				addEdge(graph, x, y); 
			}
		}
	} 

	// get the maximum value of longest path
	maximum = find_max_path(graph,n);

	// print maximum
	printf("\n%d",maximum);

	return 0;
}
